src/IECoreScene/MeshAlgoMerge.cpp

```diff
@@ -0,0 +1,297 @@
+					data2->writable().insert( data2->writable().end(), expandedData->readable().begin(), expandedData->readable().end() );
+				}
+
+				m_mesh->variables[m_name] = PrimitiveVariable( m_primVar.interpolation, data2 );
+
+				m_visitedData[data] = data2;
+			}
+		}
+
+	private :
+
+		MeshPrimitive *m_mesh;
+		const std::string m_name;
+		const PrimitiveVariable &m_primVar;
+		std::map<ConstDataPtr, DataPtr> &m_visitedData;
+
+};
+
+void merge( MeshPrimitive *a, const MeshPrimitive *b )
+{
+	const auto &vertexIdsA = a->vertexIds()->readable();
+	const auto &verticesPerFaceA = a->verticesPerFace()->readable();
+
+	const auto &vertexIdsB = b->vertexIds()->readable();
+	const auto &verticesPerFaceB = b->verticesPerFace()->readable();
+
+	IntVectorDataPtr verticesPerFaceData = new IntVectorData;
+	auto &verticesPerFace = verticesPerFaceData->writable();
+	verticesPerFace.resize( verticesPerFaceA.size() + verticesPerFaceB.size() );
+	auto it = std::copy( verticesPerFaceA.begin(), verticesPerFaceA.end(), verticesPerFace.begin() );
+	std::copy( verticesPerFaceB.begin(), verticesPerFaceB.end(), it );
+
+	IntVectorDataPtr vertexIdsData = new IntVectorData;
+	auto &vertexIds = vertexIdsData->writable();
+	vertexIds.resize( vertexIdsA.size() + vertexIdsB.size() );
+	it = std::copy( vertexIdsA.begin(), vertexIdsA.end(), vertexIds.begin() );
+	int vertexIdOffset = a->variableSize( PrimitiveVariable::Vertex );
+	std::transform( vertexIdsB.begin(), vertexIdsB.end(), it, std::bind2nd( std::plus<int>(), vertexIdOffset ) );
+
+	/// \todo: can we use setTopologyUnchecked?
+	a->setTopology( verticesPerFaceData, vertexIdsData, a->interpolation() );
+
+	/// \todo: can this be parallelized?
+	std::set<DataPtr> visitedData;
+	for( auto &pv : a->variables )
+	{
+		if( pv.second.interpolation != PrimitiveVariable::Constant )
+		{
+			IntVectorData *indices = pv.second.indices ? pv.second.indices.get() : nullptr;
+			AppendPrimVars f( a, b, pv.first, pv.second.interpolation, indices, visitedData );
+			despatchTypedData<AppendPrimVars, TypeTraits::IsVectorTypedData, DespatchTypedDataIgnoreError>( pv.second.data.get(), f );
+		}
+	}
+
+	/// \todo: can this be parallelized?
+	std::map<ConstDataPtr, DataPtr> visitedData2;
+	for( auto &pv : b->variables )
+	{
+		if( pv.second.interpolation != PrimitiveVariable::Constant )
+		{
+			PrependPrimVars f( a, pv.first, pv.second, visitedData2 );
+			despatchTypedData<PrependPrimVars, TypeTraits::IsVectorTypedData, DespatchTypedDataIgnoreError>( pv.second.data.get(), f );
+		}
+	}
+}
+
+} // namespace
+
+MeshPrimitivePtr IECoreScene::MeshAlgo::merge( const std::vector<const MeshPrimitive *> &meshes )
+{
+	if( meshes.empty() )
+	{
+		throw IECore::InvalidArgumentException( "IECoreScene::MeshAlgo::merge : No Mesh Primitives were provided." );
+	}
+
+	/// \todo: can this be parallelized?
+	MeshPrimitivePtr result = meshes[0]->copy();
+	auto it = meshes.begin() + 1;
+	for( ; it != meshes.end(); ++it )
```

Comment:
I suspect this scales poorly with increasing numbers of meshes. Rather than allocating enough memory for everything and filling it once, we're re-allocating and re-copying from the start for every single mesh added - I just discovered that this is called [Schlemiel the Painter's Algorithm](https://en.wikichip.org/wiki/schlemiel_the_painter%27s_algorithm), and it even has its own joke! Since this isn't worse than the previous option of doing the same manually via MeshMergeOp, and I suspect you want to move on to other tasks, I suggest we change the comments about parallelisation to a comment about addressing this problem instead.

Reply:
Todo updated in 5e25f85